Allow verification of SNP AttestationReport without OpenSSL

fspreiss opened the pull request:
Allows the verification of an SNP `AttestationReport` without the use of OpenSSL through a new feature called `crypto_nossl`.

In particular, with `crypto_nossl` enabled, the `certs::snp::Verifiable` trait becomes implemented in pure Rust using the `p384`, `rsa`, `x509-cert`, and `sha2` crates, that is, without the use of OpenSSL, for the following types:
* `&certs::snp::Chain`
* `&certs::snp::ca::Chain`
* `(&certs::snp::Certificate, &certs::snp::Certificate)`
* `(&certs::snp::Chain, &firmware::guest::AttestationReport)`

This addresses https://github.com/virtee/sev/issues/109.

#### Tests and Github Workflows
* Adapts the feature gate for the SNP tests in `tests/certs.rs` so that those tests are run if either `openssl` or `crypto_nossl` is enabled.
* Adds negative tests for the SNP chain and report.
* Adds a `sw-crypto_nossl` test Github workflow to also perform the various cargo tests for the `crypto_nossl` feature.
* Splits the `cargo clippy` lint Github workflow into `cargo clippy openssl` and `cargo clippy crypto_nossl`, because currently the original workflow uses `--all-features`, which is not possible any more because `openssl` and `crypto_nossl` are mutually exclusive.

#### Feature & Compatibility
* `crypto_nossl` is not enabled by default and must be enabled manually if needed.
* `crypto_nossl` is _mutually exclusive_ with the existing `openssl` feature. If both are enabled at the same time, an explicit compiler error indicates that the two are incompatible. With this, a library user who wants to verify an SNP attestation report has to make a conscious decision as to whether OpenSSL shall be used for the verification or not. If OpenSSL shall be used, `openssl` is enabled. If it is OK not to use OpenSSL for verification, then `crypto_nossl` can be enabled.
* The feature name `crypto_nossl` is chosen deliberately as it makes it very explicit for library users that OpenSSL would _not_ be used for cryptographic verification.
* The existing OpenSSL-based implementation and API remains entirely unchanged. Existing library users are unaffected by this new feature, no matter if they currently have `openssl` enabled or not.

#### Misc
Removes the `pub use crate::firmware::host::Firmware;` re-export in `src/firmware/host/types/sev.rs` because this triggers an `error: unused import: crate::firmware::host::Firmware` in `src/firmware/host/types/sev.rs:6:9` in the clippy jobs. This seems to be a new clippy lint introduced in `1.75.0-nightly`, which is currently used by the Github workflow.

larrydewey commented:
This is an interesting proposal. One of my hesitations with using the pure rust crypto libraries is that they have not been independently audited. Where there are some enterprise entities requiring FIPS certification, openssl made sense. How would you feel about adding in a separate conditional feature flag to opt for the pure rust libraries? @tylerfanelli and @DGonzalezVillal  any feelings about this?

DGonzalezVillal commented:
I understand wanting to be able to use the library without openssl, but then again we picked openssl because it's FIPS certified which a lot of people want/need. I think a conditional feature flag could be a good option to provide both solutions.

fspreiss commented:
Thanks for your comments, @larrydewey and @DGonzalezVillal, and for pointing out that some users may prefer a FIPS-certified verification library.

Based on your inputs, I adapted my proposal so that existing code isn't functionally changed, but additional code for verifying an SNP `AttestationReport` is added and gated with `#[cfg(not(feature = "openssl"))]` (note the `not`).

With this, an SNP report can be verified no matter if the optional, non-default `openssl` feature is enabled or not: if the `openssl` feature is enabled, OpenSSL is used. Otherwise, the pure-Rust implementation kicks in. This means all existing usages of the crate that have the `openssl` feature already enabled will continue to function in exactly the same way as before. Those who used the crate without the `openssl` feature so far, anyway didn't have the possibility to verify SNP reports, so also for them nothing changes, except that now they actually could verify SNP reports if they wanted to. Also, with this mechanism, there is no need for an additional feature-flag (to control if the pure-Rust implementation shall be used or not).

What do you think about this approach?

larrydewey commented:
> Thanks for your comments, @larrydewey and @DGonzalezVillal, and for pointing out that some users may prefer a FIPS-certified verification library.

No problem! Thank you for starting the PR!

> Based on your inputs, I adapted my proposal so that existing code isn't functionally changed, but additional code for verifying an SNP `AttestationReport` is added and gated with `#[cfg(not(feature = "openssl"))]` (note the `not`).
>
> With this, an SNP report can be verified no matter if the optional, non-default `openssl` feature is enabled or not: if the `openssl` feature is enabled, OpenSSL is used. Otherwise, the pure-Rust implementation kicks in. This means all existing usages of the crate that have the `openssl` feature already enabled will continue to function in exactly the same way as before. Those who used the crate without the `openssl` feature so far, anyway didn't have the possibility to verify SNP reports, so also for them nothing changes, except that now they actually could verify SNP reports if they wanted to. Also, with this mechanism, there is no need for an additional feature-flag (to control if the pure-Rust implementation shall be used or not).
> 
> What do you think about this approach?

I definitely like the idea of introducing the pure Rust solution for the cryptographic functionality, but I am a little hesitant to make it the default behavior of the application. My hesitation comes because I would be worried about someone setting the `openssl` feature flag, but then using the application thinking they are utilizing the library because the functionality is available. How would you feel about gating the functionality behind a new feature flag, and `or`-ing the logic so that only one of the options is viable at a time?

fspreiss commented:
> How would you feel about gating the functionality behind a new feature flag, and or-ing the logic so that only one of the options is viable at a time?

If I understand you correctly, you would like the new feature and `openssl` to be _mutually exclusive_. This is definitely doable. Note, however, that this is somewhat discouraged by the Cargo Book: see [this](https://doc.rust-lang.org/cargo/reference/features.html#mutually-exclusive-features).

To see how this would look like, the PR is now adapted so that the newly added code is gated with a new feature. It is called `pure_rust` for now: I'm not happy with the name, but the alternatives I could think of (e.g., `pure`, `verify`, `no_openssl`, etc.) also didn't seem much better. Maybe you have an idea. The PR description is not adapted yet. I'll adapt it once everyone is on board.

As recommended in the Rust book, I added the following:
```
#[cfg(all(feature = "openssl", feature = "purerust"))]
compile_error!("feature \"openssl\" and feature \"purerust\" cannot be enabled at the same time");
```
Note that in either case (that is, with or without the above compiler_error!), there are at least the following two compiler errors if both features are enabled at the same time: (1) `error[E0252]: the name Certificate is defined multiple times` and (2) `error[E0119]: conflicting implementations of trait certs::snp::Verifiable ...`. The additional, explicit compiler error just makes it clear what the problem is.

> My hesitation comes because I would be worried about someone setting the openssl feature flag, but then using the application thinking they are utilizing the library because the functionality is available.

What do you mean with "library" and with "is available" here? Do I understand you correctly that you are worried that someone setting the openssl feature flag could think that despite setting the openssl flag, the code would still not use OpenSSL for performing the verification, but the "pure rust" verification because this pure rust verification is also available without the openssl feature?

If yes, I'd say a simple and effective way to address this is through documentation. The README.md could say that by default the verification of an (SNP) chain and report is done purely in Rust, and that if someone wants this to be done by (the FIPS-certified) OpenSSL, the `openssl` feature has to be enabled. I'd say it is reasonable to assume that the documentation is consulted about the meaning of a feature before explicitly enabling this feature, especially if it is not enabled by default (which is the case here).
Another reason to make the pure-rust way the default is that the code becomes simpler and thus easier maintainable. For example, instead of being able to remove many `#[cfg(feature = "openssl")]`, with the new feature, those are now kept and turned into `#[cfg(any(feature = "openssl", feature = "pure_rust"))]`.

In any case, one argument that definitely speaks _for_ introducing a feature flag is that there is no good reason that someone who wants the verification done with OpenSSL has to pull in the various crates that are needed to do that (`p384`, `rsa`, `x509-cert`, etc.).

Let me know what you think.

fspreiss commented:
Update on the above comment: The `lint / cargo clippy` CI job is now failing because it enables `--all-features` which is of course not possible any more, so this would have to be adapted/fixed.

tylerfanelli commented:
Thanks @fspreiss, would you mind squashing the commits and adding a `Signed-off-by` line in the commit message?

tylerfanelli commented:
> > How would you feel about gating the functionality behind a new feature flag, and or-ing the logic so that only one of the options is viable at a time?
> 
> If I understand you correctly, you would like the new feature and `openssl` to be _mutually exclusive_. This is definitely doable. Note, however, that this is somewhat discouraged by the Cargo Book: see [this](https://doc.rust-lang.org/cargo/reference/features.html#mutually-exclusive-features).
> 
> To see how this would look like, the PR is now adapted so that the newly added code is gated with a new feature. It is called `pure_rust` for now: I'm not happy with the name, but the alternatives I could think of (e.g., `pure`, `verify`, `no_openssl`, etc.) also didn't seem much better. Maybe you have an idea. The PR description is not adapted yet. I'll adapt it once everyone is on board.

I much prefer something like `crypto_nossl` or something similar.

> 
> As recommended in the Rust book, I added the following:
> 
> ```
> #[cfg(all(feature = "openssl", feature = "purerust"))]
> compile_error!("feature \"openssl\" and feature \"purerust\" cannot be enabled at the same time");
> ```
> 
> Note that in either case (that is, with or without the above compiler_error!), there are at least the following two compiler errors if both features are enabled at the same time: (1) `error[E0252]: the name Certificate is defined multiple times` and (2) `error[E0119]: conflicting implementations of trait certs::snp::Verifiable ...`. The additional, explicit compiler error just makes it clear what the problem is.
> 
> > My hesitation comes because I would be worried about someone setting the openssl feature flag, but then using the application thinking they are utilizing the library because the functionality is available.

I have this same concern. Right now, if a user would like to verify certificates without the `openssl` feature, it will simply fail and they will know to add the feature. With this, they will be verifying and potentially thinking they are using OpenSSL, but are not.

> 
> What do you mean with "library" and with "is available" here? Do I understand you correctly that you are worried that someone setting the openssl feature flag could think that despite setting the openssl flag, the code would still not use OpenSSL for performing the verification, but the "pure rust" verification because this pure rust verification is also available without the openssl feature?

I believe Larry is indicating my above point, where they will be unknowingly using the rust crypto solution instead of something like OpenSSL (which they will have _much_ more trust in).

> 
> If yes, I'd say a simple and effective way to address this is through documentation. The README.md could say that by default the verification of an (SNP) chain and report is done purely in Rust, and that if someone wants this to be done by (the FIPS-certified) OpenSSL, the `openssl` feature has to be enabled.

I disagree. I think that both should be optional features that need to be explicitly enabled.

> I'd say it is reasonable to assume that the documentation is consulted about the meaning of a feature before explicitly enabling this feature, especially if it is not enabled by default (which is the case here). Another reason to make the pure-rust way the default is that the code becomes simpler and thus easier maintainable. For example, instead of being able to remove many `#[cfg(feature = "openssl")]`, with the new feature, those are now kept and turned into `#[cfg(any(feature = "openssl", feature = "pure_rust"))]`.
> 
> In any case, one argument that definitely speaks _for_ introducing a feature flag is that there is no good reason that someone who wants the verification done with OpenSSL has to pull in the various crates that are needed to do that (`p384`, `rsa`, `x509-cert`, etc.).
> 
> Let me know what you think.

While I like this, I think that something like OpenSSL is watched and maintained very carefully, and that security-conscious users will definitely take issue with mistakenly using non-OpenSSL for crypto-related functionalty. Thus, I think that this should be a non-default feature.

fspreiss commented:
@tylerfanelli, thanks for your thoughts! What you write makes sense, and the new feature flag definitely seems the way to go here, so as to avoid any possible confusion regarding whether OpenSSL is used or not. With this in mind, the name `no_openssl` seems to make sense as feature name, so I renamed accordingly. I updated the PR description, fixed the CI issues, and squashed the commits with a sign-off. Please take a look.

DGonzalezVillal commented:
> I think that both should be optional features that need to be explicitly enabled.

I think this is the best solution. Making them mutually exclusive seems to be adding too much of a complicated layer.  If we make it so that the user has to at least enable 1, then the chance of confusion should be reduced, unless they for some reason decide to enable both at the same time. But I feel the library should still compile in that case.

DGonzalezVillal reviewed (approved):
code lgtm

DGonzalezVillal commented on src/lib.rs:
```diff
@@ -46,6 +46,9 @@
 #![allow(clippy::unreadable_literal)]
 
+#[cfg(all(feature = "openssl", feature = "no_openssl"))]
+compile_error!("feature \"openssl\" and feature \"no_openssl\" cannot be enabled at the same time");
```
Would there be a scenario were we want the ability to have both enabled? @larrydewey Or do we best keep them separate?

DGonzalezVillal reviewed (commented):
I

larrydewey reviewed (changes requested):
This looks great! My only ask would be to change out the generalized IO Errors into meaningful library errors (using error.rs).

tylerfanelli commented:
Small change, but I would prefer `crypto_nossl` instead of `no_openssl`.. a bit clearer.

tylerfanelli commented:
Also, a note in the README explaining the crypto_nossl feature would be useful.

fspreiss commented:
> Small change, but I would prefer crypto_nossl instead of no_openssl

> Also, a note in the README explaining the crypto_nossl feature would be useful.

@tylerfanelli, Done (for both of the above).

> My only ask would be to change out the generalized IO Errors into meaningful library errors (using error.rs).

@larrydewey, I'm happy to make some errors more specific, but I'm not sure I understand which exact errors you would prefer at which locations. There are the following locations in the new code where errors are used:

1. In `impl Verifiable for (&Chain, &AttestationReport)` and `impl Verifiable for (&Certificate, &Certificate)`

   Using a non-io-Error is not possible there, because the error type is determined by the `Verifiable` trait. Changing the actual trait definition seems somewhat out of scope, no?

2. In `impl TryFrom<&Signature> for p384::ecdsa::Signature`

   I'm using `io::Error` because this exact error is also used for all the other (three) `Signature`-related `TryFrom` conversions in that (`ecdsa`) module. It seemed to make sense to stay consistent with the rest of the code. If you prefer a different/dedicated error for this specific `TryFrom`, did you mean to re-use an existing error from `errors.rs` (if yes, which one), or did you mean to introduce a new one?

3. In the various methods in `impl Certificate` (`from_pem`, `to_pem`, `from_der`, `to_der`):

   I used the exact same API (incl. the same error types) that is used for the `openssl`-based implementation of `Certificate`. This way, library users can seamlessly switch between `openssl` and `crypto_nossl` without having to adapt any user-side code. If we were to switch away from `io::Error`, this seamless switch wouldn't be possible any more. Still, the errors there could be changed if you prefer that. If yes, did you mean to re-use an existing error from `errors.rs` (if yes, which one), or did you mean to introduce new ones?

larrydewey commented:
You bring up some good points. Let's not focus on those adjustments in this PR. However, @tylerfanelli it may be worth looking at not using a `std::io::Result`, as there are potentially other Errors which may come up from this. We can discuss it on the mailing list.

tylerfanelli commented:
@fspreiss Can you address the `Cargo.toml` conflict?

fspreiss commented:
> Can you address the Cargo.toml conflict?

@tylerfanelli, Done.

I had to force-push again so that everything is in a single signed-off commit. Looks like because of this the approvals need to be given again.

DGonzalezVillal reviewed (approved):
lgtm
